wasm/src/project_system.rs

```diff
@@ -0,0 +1,272 @@
+
+#[wasm_bindgen]
+extern "C" {
+    #[wasm_bindgen(typescript_type = "(uri: string) => Promise<string | null>")]
+    pub type ReadFileCallback;
+}
+
+#[wasm_bindgen]
+extern "C" {
+    #[wasm_bindgen(typescript_type = "(uri: string) => Promise<[string, number][]>")]
+    pub type ListDirectoryCallback;
+}
+
+#[wasm_bindgen]
+extern "C" {
+    #[wasm_bindgen(
+        typescript_type = "{ excludeFiles: string[], excludeRegexes: string[], manifestDirectory: string }"
+    )]
+    pub type ManifestDescriptorObject;
+}
+
+#[wasm_bindgen]
+extern "C" {
+    #[wasm_bindgen(
+        typescript_type = "(uri: string) => Promise<{ excludeFiles: string[], excludeRegexes: string[], manifestDirectory: string }| null>"
+    )]
+    pub type GetManifestCallback;
+}
+
+#[wasm_bindgen]
+extern "C" {
+    #[wasm_bindgen(typescript_type = "[string, string][]")]
+    pub type ProjectSources;
+}
+
+impl From<ManifestDescriptorObject> for Option<ManifestDescriptor> {
+    fn from(value: ManifestDescriptorObject) -> Self {
+        get_manifest_transformer(value.obj, Default::default())
+    }
+}
+
+/// This macro produces a function that calls an async JS function, awaits it, and then applies a function to the resulting value.
+/// Ultimately, it returns a function that accepts a String and returns a Rust future that represents a JS Promise. Awaiting that
+/// Rust future will await the resolution of the promise.
+/// The name of this macro should be read like "convert a JS promise into an async rust function with this mapping function"
+macro_rules! into_async_rust_fn_with {
+    ($js_async_fn: ident, $map_result: expr) => {{
+        use crate::project_system::{map_js_promise, to_js_function};
+        use std::future::Future;
+        use std::pin::Pin;
+        use wasm_bindgen::JsValue;
+        use wasm_bindgen_futures::JsFuture;
+
+        let $js_async_fn = to_js_function($js_async_fn, stringify!($js_async_fn));
+
+        let $js_async_fn = move |input: String| {
+            let path = JsValue::from_str(&input);
+            let res: js_sys::Promise = $js_async_fn
+                .call1(&JsValue::NULL, &path)
+                .expect("callback should succeed")
+                .into();
+
+            let res: JsFuture = res.into();
+
+            Box::pin(map_js_promise(res, move |x| $map_result(x, input.clone())))
+                as Pin<Box<dyn Future<Output = _> + 'static>>
+        };
+        $js_async_fn
+    }};
+}
+
+pub(crate) async fn map_js_promise<F, T>(res: JsFuture, func: F) -> T
+where
+    F: Fn(JsValue) -> T,
+{
+    let res = res.await.expect("js future shouldn't throw an exception");
+    func(res)
+}
+
+pub(crate) fn to_js_function(val: JsValue, help_text_panic: &'static str) -> js_sys::Function {
+    let js_ty = val.js_typeof();
+    TryInto::<js_sys::Function>::try_into(val).unwrap_or_else(|_| {
+        panic!(
+            "expected a valid JS function ({help_text_panic}), received {:?}",
+            js_ty
+        )
+    })
+}
```

Comment:
Newer clippy gets mad about this line because the conversion is actually infallible. If you want to check the type, you should probably use `.is_function()` and panic if the result is false, then just return the `Into` conversion.


```suggestion
    assert!(
        val.is_function(),
        "expected a valid JS function ({help_text_panic}), received {:?}",
        js_ty
    );
    Into::<js_sys::Function>::into(val)
```

Reply:
whoops, looks like something in the formatting here got messed up, so github isn't going to put the suggestion in the right spot. Might have to copy paste manually.

Reply:
will be addressed in follow up PRs on main